Entity factory article TRNG-952 (#134)
Innoactive/Creator: .Documentation/articles/developer/07-run-a-course.md

> # Create a Course

Reviewer: I would make it a separate chapter and name it to "Create a Course Programmatically".

Reviewer: By that I mean a separate file, 08-create-a-course-programmatically.md

Don't forget to update TOC and the rest of the file names.

## Entity Factory

> The `EntityFactory` is a helper class that allows you to dynamically create basic elements from a course. It is part of the Innoactive Creator's API.

Reviewer: ```suggestion
The `EntityFactory` is a helper class that allows you to dynamically create basic elements of a course.
```

> It supports the creation of steps (`IStep`), transitions (`ITransition`), chapters (`IChapter`), and courses (`ICourse`). A course can be edited as much as possible only if it is not being played.

Reviewer: ```suggestion
It supports the creation of steps (`IStep`), transitions (`ITransition`), chapters (`IChapter`), and courses (`ICourse`). You can modify the course as much as you want until the application starts executing it.
```

Reviewer: Is it actually true? Can't you add new steps to the course as it plays out?

Author: Actually, I am not sure, I thought that even if it would be possible it might mess with the system, but let me try it out.

Author: Apparently it is possible to edit the course while it is been playing, at least no errors are thrown, but I found a lot of difficulties for adding more steps, for instance after adding only one step the course executes that new dynamic step but it gets stuck in the Activation phase, not sure if it has to do with the course being played on runtime or it actually will happen even if the course is edited or created dynamically 😕

Additionally, the `EntityFactory` provides a built-in post-processing stack for the recently built elements.

## Usage

> It is possible to easely create a course element with a single line of code, adding corresponding element to where it belongs also takes a single line, which makes courses creation very flexible and easy to execute:

Reviewer: ```suggestion
It takes only a line of code to create a course element, and one more line to assign it to the course:
```

```csharp
using UnityEngine;
using Innoactive.Creator.Core;

// Unity components' classes should have the same names as their files.
public class TrainingCourseBuilder : MonoBehaviour
{
    public ICourse CreateDefaultCourse()
    {
        // Create a new ICourse.
        ICourse course = EntityFactory.CreateCourse("New Course");
        
        // Create a new IChapter.
        IChapter chapter = EntityFactory.CreateChapter("Chapter 1");
        
        // Create a new IStep.
        IStep step = EntityFactory.CreateStep("New Step");
        
        // Create a new ITransition.
        ITransition transition = EntityFactory.CreateTransition();
        

>         // Adding an ITransition to an IStep.

Reviewer: ```suggestion
        // Add the transition to the step.
```

        step.Data.Transitions.Data.Transitions.Add(transition);
        

>         // Adding an IStep to an IChapter.

Reviewer: ```suggestion
        // Add the step to the chapter.
```

        chapter.Data.Steps.Add(step);
        

>         // Adding an IChapter to an ICourse.

Reviewer: ```suggestion
        // Add the chapter to the course.
```

        course.Data.Chapters.Add(chapter);

        return course;
    }
}
```

> By default, all `Istep` and `IChapter` created with the `EntityFactory` already contain a default `ITransition` and `IChapter` respectively. For reference purposes, in the example above, we intentionally show how to create and add them to a course.

Reviewer: ```suggestion
By default, all steps and chapters created with the `EntityFactory` already contain a default transition and a chapter. we explicitly show how to create and add them to a course in the example above for reference purposes.
```

Reviewer: Doesn't make much sense for me. Should it be "all steps and courses already contain a default transition and a chapter"?

Author: Yes you are right, it should be steps and courses.

## Entity PostProcessing

> In the case where it is required to execute certain actions to every new entity, this can be achieved by extending the `EntityPostProcessing <T>` class, T corresponds to the type of the element we aim to post-process.

Reviewer: ```suggestion
You might want to execute some actions over freshly created entities. You can do it by extending the the `EntityPostProcessing <T>` class, where T corresponds to the type of the element you want to post-process.
```

> The post-processing stack is automatically fired by the `EntityFactory` every time it creates a course element. 

Reviewer: Every time the `EntityFactory` creates a course element, it executes the post-processing stack of actions.
